src/dewey/indexing.clj

```diff
@@ -1,72 +1,77 @@
 (ns dewey.indexing
   "This is the logic for making changes to search index."
-  (:require [clojurewerkz.elastisch.query :as es-query]
-            [clojurewerkz.elastisch.rest.document :as es-doc]
-            [clojurewerkz.elastisch.rest :as rest]
+  (:use [slingshot.slingshot :only [try+]])
+  (:require [qbits.spandex :as s]
             [clojure-commons.file-utils :as file]
             [dewey.doc-prep :as prep]
             [dewey.config :as cfg]
-            [dewey.entity :as entity])
-  (:import [java.util Map]
-           [clojure.lang Keyword]))
+            [dewey.entity :as entity]
+            [clojure.tools.logging :as log]))
 
 
-(def ^{:private true} collection-type "folder")
-(def ^{:private true} data-object-type "file")
+(defn index-doc
+  [es doc]
+  (s/request es {:url
+                 [(cfg/es-index) :_doc (str (:id doc))]
+                 :method :put
+                 :headers {"Content-Type" "application/json"}
+                 :body doc}))
```

Comment:
This is the spot where we could be replacing whole documents which needs to be avoided/modified. Since it's using a `PUT` to the `:index/_doc/:id` endpoint, this is the [Index API](https://www.elastic.co/guide/en/elasticsearch/reference/7.16/docs-index_.html).

I think what needs to happen here is:

* change this to use the `_update` API that `update-doc` does, at which point it's sending a partial document and should perform an update
* set the [`doc_as_upsert`](https://www.elastic.co/guide/en/elasticsearch/reference/7.16/docs-update.html#doc_as_upsert) option so that if the document doesn't already exist, it will create it despite the use of the `_update` endpoint
* test carefully -- in particular: we should manually add a `metadata.cyverse` section to a document and then send an update that sends the whole document including a `metadata.irods` section, but not including a `metadata.cyverse` section to ensure it doesn't touch that; at the other end, we should send some documents that change things within arrays (so, irods metadata or permissions come to mind), and make sure that the old values are in fact replaced with the new ones as they should be

If both parts of the last point work correctly, then we're probably good. If they don't, then what we probably need to do is actually adapt the indexing to use a scripted update and correctly update each subpart of the document. We'd still want to send the whole document as an `upsert`, but without the `doc_as_upsert` stuff in this case.

Hopefully this makes sense, but definitely ask any questions you have!

Reply:
The index_doc function only seems to run in cases of an entity not being indexed. It seems that Dewey was built around the update-or-index function which checks whether an entity should be updated or indexed based off whether it is already indexed, thus preventing overwriting. 

For clarity, we could use the _create endpoint here which will throw an error if trying to create an already existing document. Per the Elastic 7.10 docs ["You can index a new JSON document with the _doc or _create resource. Using _create guarantees that the document is only indexed if it does not already exist. To update an existing document, you must use the _doc resource."](https://www.elastic.co/guide/en/elasticsearch/reference/7.10/docs-index_.html)

Reply:
Ah, right. I think I may have conflated in some other services and thought that there were cases where dewey would update the full document despite it already existing. Using `_create` seems like a nice change for safety -- a lot of the time there's a burst of messages for one object when it's initially created, and that'd prevent it overwriting in the race condition -- but not critical either.